os/drivers/audio/syu645b.c

```diff
@@ -0,0 +1,1132 @@
+
+#include <tinyara/config.h>
+#include <sys/types.h>
+#include <sys/ioctl.h>
+#include <stdint.h>
+#include <stdbool.h>
+#include <stdio.h>
+#include <unistd.h>
+#include <semaphore.h>
+#include <queue.h>
+#include <errno.h>
+#include <fixedmath.h>
+#include <debug.h>
+#include <tinyara/kmalloc.h>
+#include <tinyara/fs/ioctl.h>
+#include <tinyara/audio/i2s.h>
+#include <tinyara/audio/audio.h>
+#include <tinyara/audio/syu645b.h>
+#include <tinyara/math.h>
+#include <math.h>
+#include <tinyara/i2c.h>
+
+#include "syu645b.h"
+#include "syu645bscripts.h"
+
+#define SYU645B_I2S_TIMEOUT_MS 200
+
+/* Default configuration values */
+
+/*
+ * It's good to match the buffer size with i2s DMA page size
+ */
+#ifndef CONFIG_SYU645B_BUFFER_SIZE
+#define CONFIG_SYU645B_BUFFER_SIZE       2048
+#endif
+
+#ifndef CONFIG_SYU645B_NUM_BUFFERS
+#define CONFIG_SYU645B_NUM_BUFFERS       4
+#endif
+
+/****************************************************************************
+ * Private Data
+ ****************************************************************************/
+
+static const struct audio_ops_s g_audioops = {
+        .getcaps = syu645b_getcaps,             /* getcaps        */
+        .configure = syu645b_configure,         /* configure      */
+        .shutdown = syu645b_shutdown,           /* shutdown       */
+        .start = syu645b_start,                 /* start          */
+#ifndef CONFIG_AUDIO_EXCLUDE_STOP
+        .stop = syu645b_stop,                   /* stop           */
+#endif
+#ifndef CONFIG_AUDIO_EXCLUDE_PAUSE_RESUME
+        .pause = syu645b_pause,                 /* pause          */
+        .resume = syu645b_resume,               /* resume         */
+#endif
+        .allocbuffer = NULL,                    /* allocbuffer    */
+        .freebuffer = NULL,                     /* freebuffer     */
+        .enqueuebuffer = syu645b_enqueuebuffer, /* enqueue_buffer */
+        .cancelbuffer = syu645b_cancelbuffer,   /* cancel_buffer  */
+        .ioctl = syu645b_ioctl,                 /* ioctl          */
+        .read = NULL,                           /* read           */
+        .write = NULL,                          /* write          */
+        .reserve = syu645b_reserve,             /* reserve        */
+        .release = syu645b_release,             /* release        */
+};
+
+static int syu645b_writereg_2byte(FAR struct syu645b_dev_s *priv, uint8_t regaddr, uint8_t regval)
+{
+        int ret;
+        uint8_t reg[2];
+        FAR struct i2c_dev_s *dev = priv->i2c;
+        FAR struct i2c_config_s *syu645b_i2c_config = &(priv->lower->i2c_config);
+
+        reg[0] = regaddr;
+        reg[1] = regval;
+
+        ret = i2c_write(dev, syu645b_i2c_config, (uint8_t *)reg, 2);
+        if (ret < 0) {
+		auddbg("Error, cannot write reg 0x%x/0x%x\n", regaddr, regval);
+        }
+        return ret;
+}
+
+static int syu645b_writereg_3byte(FAR struct syu645b_dev_s *priv, uint8_t regaddr, uint8_t regval[4])
```

Comment:
all of this looks redundant, I think better way is to create a function like syu645b_writereg_nbyte and pass the length as argument.
```
static int syu645b_writereg_nbyte(FAR struct syu645b_dev_s *priv, uint8_t regaddr, uint8_t* regval, size_t nbytes);
```